pkg/sql/opt/optbuilder/distinct.go

```diff
@@ -208,15 +212,20 @@ func (b *Builder) analyzeDistinctOnArgs(
 	b.semaCtx.Properties.Require(exprKindDistinctOn.String(), tree.RejectGenerators)
 	inScope.context = exprKindDistinctOn
 
-	for i := range distinctOn {
+	for i := range d {
 		b.analyzeExtraArgument(
-			distinctOn[i],
+			&d[i],
 			inScope,
 			projectionsScope,
 			distinctOnScope,
 			true, /* nullsDefaultOrder */
 		)
 	}
+
+	if b.insideFuncDef {
+		*distinctOn = d
```

Comment:
I don't understand the need for this - the length of `d` or its underlying array are not mutated above, are they?